libraries/AP_AHRS/AP_AHRS_Logging.cpp

```diff
@@ -85,16 +93,27 @@ void AP_AHRS::Write_POS() const
     if (!get_location(loc)) {
         return;
     }
-    float home, origin;
-    AP::ahrs().get_relative_position_D_home(home);
+    const auto nanf = AP::logger().quiet_nanf();
+    float home_pos_down;
+    if (!AP::ahrs().get_relative_position_D_home(home_pos_down)) {
+        home_pos_down = nanf;
+    }
+    float origin_pos_down;
+    if (!get_relative_position_D_origin(origin_pos_down)) {
+        origin_pos_down = nanf;
+    }
+    float alt;
+    if (!loc.get_alt_m(Location::AltFrame::ABSOLUTE, alt)) {
+        alt = nanf;
+    }
     const struct log_POS pkt{
         LOG_PACKET_HEADER_INIT(LOG_POS_MSG),
         time_us        : AP_HAL::micros64(),
         lat            : loc.lat,
         lng            : loc.lng,
-        alt            : loc.alt*1.0e-2f,
-        rel_home_alt   : -home,
-        rel_origin_alt : get_relative_position_D_origin(origin) ? -origin : AP::logger().quiet_nanf(),
+        alt            : alt,
+        rel_home_alt   : -home_pos_down,   // down->up
```

Comment:
Will negating a nan result in internal errors? I'm fairly sure its fine we only check on things like constrain.

Reply:
Good question.

I've restructured the code to avoid the question :-)